src/presign/participant.rs

```diff
@@ -85,6 +85,37 @@ pub enum Status {
     TerminatedSuccessfully,
 }
 
+/// PresignContext includes shared parameters in
+/// [`SharedContext`] and public parameters as
+/// defined in [`AuxInfoPublic`].
+pub struct PresignContext {
+    shared_context: SharedContext,
+    auxinfo_public: Vec<AuxInfoPublic>,
+}
+
+impl ProofContext for PresignContext {
+    fn as_bytes(&self) -> Result<Vec<u8>> {
+        Ok([
+            self.shared_context.as_bytes()?,
+            bincode::serialize(&self.auxinfo_public)
+                .map_err(|_| InternalError::InternalInvariantFailed)?,
+        ]
+        .concat())
+    }
+}
+
+impl PresignContext {
+    pub(crate) fn collect<P: InnerProtocolParticipant>(
+        auxinfo_public: &[AuxInfoPublic],
+        p: &P,
```

Comment:
todo: This function doesn't have to be generic over protocol participants since we're only using it for presign! You can take a `participant: &PresignParticipant` directly and access `participant.all_auxinfo_public()` within this function instead of passing it as a parameter.